src/conv.jl

```diff
@@ -169,25 +173,106 @@ end
 # im2col + GEMM.  Do so in a loop, here:
+
+# These are the GEMM types we will accelerate with `im2col`
+const G = Union{[x[2] for x in gemm_datatype_mappings]...}
+
 for (front_name, backend) in (
         # This maps from public, front-facing name, to internal backend name
         :conv                   => :im2col,
-        :∇conv_data             => :im2col,
-        :∇conv_filter           => :im2col,
+    )
+
+    # We only define 3d conv primitives, we reshape lower down to get 1d and 2d convolution
+    @eval begin
+        # im2col-accelerated function forwarding definition
+        function $(Symbol("$(front_name)!"))(
+                        out::AbstractArray{T,5}, in1::AbstractArray{T,5},
+                        in2::AbstractArray{T,5}, cdims::C; kwargs...) where {T <: $G, C <: ConvDims}
+
+            x_cs = Iterators.partition(1:size(in1, 4),
+                                       channels_in(cdims) ÷ groupcount(cdims))
+            w_cs = Iterators.partition(1:size(in2, 5),
+                                       channels_out(cdims) ÷ groupcount(cdims))
+            cdims2 = basetype(C)(cdims,
+                                 G = 1,
+                                 C_in = channels_in(cdims) ÷ groupcount(cdims),
+                                 C_out = channels_out(cdims) ÷ groupcount(cdims))
+            
+            Threads.@sync for (xc, wc) in zip(x_cs, w_cs)
+              x = @view in1[ntuple(i -> i == 4 ? xc : Colon(), 5)...]
+              w = @view in2[ntuple(i -> i == 5 ? wc : Colon(), 5)...]
+              y = @view out[ntuple(i -> i == 4 ? wc : Colon(), 5)...]
+              Threads.@spawn $(Symbol("$(front_name)_$(backend)!"))(y, x, w, cdims2; kwargs...)
+            end
+
+           return out
+        end
+    end
+end
+
+# im2col-accelerated function forwarding definition
+function ∇conv_data!(out::AbstractArray{T,5}, in1::AbstractArray{T,5},
+                     in2::AbstractArray{T,5}, cdims::C; kwargs...) where {T <: G, C <: ConvDims}
+
+    dx_cs = Iterators.partition(1:size(out, 4),
+                                channels_in(cdims) ÷ groupcount(cdims))
+    w_cs = Iterators.partition(1:size(in2, 5),
+                               channels_out(cdims) ÷ groupcount(cdims))
+    dy_cs = Iterators.partition(1:size(in1, 4),
+                                channels_out(cdims) ÷ groupcount(cdims))
+    cdims2 = basetype(C)(cdims,
+                         G = 1,
+                         C_in = channels_in(cdims) ÷ groupcount(cdims),
+                         C_out = channels_out(cdims) ÷ groupcount(cdims))
+
+    Threads.@sync for (xc, yc, wc) in zip(dx_cs, dy_cs, w_cs)
+      dxv = @view out[ntuple(i -> i == 4 ? xc : Colon(), 5)...]
+      dyv = @view in1[ntuple(i -> i == 4 ? yc : Colon(), 5)...]
+      wv = @view in2[ntuple(i -> i == 5  ? wc : Colon(), 5)...]
+      Threads.@spawn ∇conv_data_im2col!(dxv, dyv, wv, cdims2; kwargs...)
+    end
```

Comment:
does the spawn approach allow for composable threading? Ideally each `conv` call on the slices should be multithreaded as well if the machine has enough threads

Reply:
This is what I see, which suggests the impact on the G = 1 case is tractable. I have answered this in a [regular comment](https://github.com/FluxML/NNlib.jl/pull/289#issuecomment-880602522) since GitHub didn't show me the reply pane for some reason.

```Julia
julia> x′ = rand(Float32, 28, 28, 20, 2);

julia> w′ = rand(Float32, 3, 3, 20, 15);

julia> cdims = DenseConvDims(x′, w′);

julia> @btime conv($x′, $w′, $cdims);
  230.201 μs (23 allocations: 555.95 KiB) # master; -t1
  243.761 μs (50 allocations: 557.77 KiB) # PR; -t1

  205.723 μs (28 allocations: 1.01 MiB) # master; -t2
  212.537 μs (47 allocations: 1.01 MiB)  # PR; -t2

julia> w′ = rand(Float32, 3, 3, 3, 6);

julia> x′ = rand(Float32, 28, 28, 3, 2);

julia> cdims = DenseConvDims(x′, w′);

julia> @btime conv($x′, $w′, $cdims);
  47.205 μs (23 allocations: 104.39 KiB) # master; -t1
  55.600 μs (50 allocations: 106.20 KiB) # PR; -t1

  35.217 μs (27 allocations: 176.05 KiB) # master; -t2
  40.850 μs (47 allocations: 177.14 KiB) # PR; -t2
```

Reply:
A little regression, but definitely worth it for this important feature. It will only affect cpu conv, so no big deal